0000-notes-on-oer-encoding/0000-notes-on-oer-encoding.md

```diff
@@ -0,0 +1,190 @@
+
+Using a formalized language like ASN.1 to describe one's protocol means that one can use automated tools to verify the consistency and completeness of one's specification. This is a very valuable tool. It also means that one can refer to existing specifications which already cover many edge cases. Otherwise, each protocol would have to answer all possible edge cases itself, which quickly becomes tedious.
+
+ASN.1 is by far the most widely used formalized language for defining protocols. As an ISO and ITU standard, it is used in telecommunications protocols, cryptography standards and many more.
+
+The Interledger team chose ASN.1 because of it's long history and robust ecosystem. It may not be the most modern of standards, but it is extremely powerful and comes with huge amounts of implementation experience. Also, there exist high-quality tools for protocol designers.
+
+## CANONICAL-OER
+
+### What is OER/CANONICAL-OER?
+
+The Octet Encoding Rules (OER) are a set of simple byte-aligned encoding rules for ASN.1 types.
+
+CANONICAL-OER narrows BASIC-OER to a single canonical encoding.
+
+### Why CANONICAL-OER?
+
+For Interledger, we wanted to use the power of ASN.1 in order to design and validate our protocols and to avoid any ambiguity for implementors. However, we also wanted it to be easy to implement Interledger encodings, without having to first learn a bunch of ASN.1 and related technology. OER gives us the best of both worlds, because its encodings are extremely simple and can easily be described and implemented without referencing ASN.1 or OER. Yet, we can use standard tools to generate test data and validate our implementations. And we can refer to the ASN.1/OER specs to resolve any ambiguities that are discovered later.
+
+OER is designed to strike a better compromise between simplicity, speed and size than previous ASN.1 encoding rules. All OER types are byte-aligned, which makes them easier to serialize and parse in most programming languages than Packed Encoding Rules (PER) and also faster on most platforms.
+
+We chose CANONICAL-OER instead of BASIC-OER because we wanted a canonical encoding format so that Interledger encodings could be used for cryptographic operations like hashing and signing.
+
+## Basics
+
+### Octet strings
+
+All Interledger formats are byte-aligned messages and should be passed as octet strings, i.e. sequences of bytes. The correct type for this would be a `Buffer` in JavaScript, `byte[]` in Java, `bytes` in Python, etc.
+
+### Extensibility
+
+For most Interledger formats, we require that implementations MUST ignore extra bytes at the end of the message. This behavior allows additional fields to be added in the future that are automatically ignored by existing implementations.
```

Comment:
end of the message == end of the message inside the envelope. Extra bytes after the ILP packet would be considered the next message, right?

Reply:
I was considering talking about ILP packets here, but this is the general OER guide, so I opted not to. I think anything that's specific to one protocol should go into that protocol's spec.

Reply:
I think this doc should talk about SEQUENCES and how these are used to define most compound types in Interledger protocols. 

For extensibility these will often contain an extensibility byte as the last element in the sequence. This will be `00` (indicating a length of zero) for most data sequences but where a non-zero value is encountered this should be interpreted as a length indicator and the subsequent `n` bytes (where `n` is the length specified by the length indicator) can be ignored.

Reply:
It will definitely help to show sample of entire stream.
e.g. sample ASN.1 code and how it is encoded by OER.